allen-cell-animated/timelapse-colorizer
# feature: Load features asynchronously

Problem
=======
Part 3/3 to close #20, handling large numbers of features! This solves a big performance problem I was seeing on initial load where the dataset would stall.

Before, the Dataset would **fetch the feature data for EVERY feature** on load. This would cause it to block for up to 10-15 seconds on the latest NucMorph datasets. This change now allows it to only load in feature data on demand, making initial startup MUCH faster.

However, there are still some interactions that could use loading spinners + better feedback when loading is happening, which I've described in #390.

This builds off of the previous changes to the `DataCache` that allows it to also store feature data and protect certain keys from eviction.

*Estimated size: large, ~40 minutes*

Solution
========

- Dataset now asynchronously loads and stores feature data in a `DataCache`.
- When a dataset is opened (loaded), only the first feature will be fetched and loaded. Other features will be loaded on demand.
- The `Viewer` now stores the selected feature's data (with the `featureData` state) instead of just the feature's key. This allows it to do synchronous data lookups.
- Updates other components to use async logic when handling feature data fetching.

## Type of change
* New feature (non-breaking change which adds functionality)

Steps to Verify:
----------------
1. Open the dataset preview link in an incognito/private browser from the main build: https://allen-cell-animated.github.io/timelapse-colorizer/viewer?collection=https%3A%2F%2Fallencell.s3.amazonaws.com%2Faics%2Fnuc-morph-dataset%2Ftimelapse_feature_explorer_datasets%2Fexploratory_dataset%2Fcollection.json&dataset=Large (this takes about 10 seconds on my computer)
3. Close the previous browser page. Open the PR preview link in a new incognito/private browser, which should load much faster: https://allen-cell-animated.github.io/timelapse-colorizer/pr-preview/pr-395/viewer?collection=https%3A%2F%2Fallencell.s3.amazonaws.com%2Faics%2Fnuc-morph-dataset%2Ftimelapse_feature_explorer_datasets%2Fexploratory_dataset%2Fcollection.json&dataset=Large
4. Click around both. Adding/removing filters or switching features will cause new feature data to be loaded in.

Screenshots (optional):
-----------------------
Show-n-tell images/animations here

Keyfiles (delete if not relevant):
-----------------------
1. `Dataset.ts`
5. `Viewer.tsx`


Thanks for contributing!
